### PR TITLE
Add billing to device lifecycle

### PR DESCRIPTION
**Pull request by knolleary**
Part of #782 

This adds a billing step to the device lifecycle.

If `billing.stripe.device_product/device_price` are set, then whenever a device is added/removed from team, the billing subscription item is updated for the team to reflect how many devices are in the team.

If the team type includes a `deviceFreeAllocation` property, this is taken into account when identifying how many 'billable' devices a team has.

This PR does not update the UI - that will come in the final PR for this story.

### CODE REVIEW

File: test/unit/forge/ee/lib/billing/index_spec.js [hunk 1]
```diff
@@ -167,4 +167,141 @@ describe('Billing', function () {
             updateData.should.have.property('proration_behavior', 'always_invoice')
         })
     })
+
+    describe.only('updateTeamDeviceCount', async function () {
```
Comment (Steve-Mcl) [inline test/unit/forge/ee/lib/billing/index_spec.js:171]:
```suggestion
    describe('updateTeamDeviceCount', async function () {
```

remove `.only`

Comment (knolleary) [reply test/unit/forge/ee/lib/billing/index_spec.js:171]: 🤦🏻

File: forge/ee/lib/billing/index.js [hunk 2]
```diff
@@ -211,6 +217,51 @@ module.exports.init = async function (app) {
...
+        updateTeamDeviceCount: async (team) => {
+            const billingIds = getBillingIdsForTeam(team)
+            if (!billingIds.device.product) {
+                return
+            }
+            const subscription = await app.db.models.Subscription.byTeam(team.id)
+            if (subscription) {
+                const deviceCount = await team.deviceCount()
+                const deviceFreeAllocation = team.TeamType.getProperty('deviceFreeAllocation') || 0
+                const billableCount = Math.max(0, deviceCount - deviceFreeAllocation)
```
Comment (Steve-Mcl) [inline forge/ee/lib/billing/index.js:229]: Question: I cannot see where `deviceFreeAllocation` is ever set (different PR? Future? From the License?)

Comment (Steve-Mcl) [reply forge/ee/lib/billing/index.js:229]: Also, should `deviceFreeAllocation` be covered in tests?

**Review by Steve-Mcl — COMMENTED** [review]
1st review pass complete

File: forge/ee/lib/billing/index.js (hunk 2 above)
Comment (knolleary) [reply forge/ee/lib/billing/index.js:229]: `deviceFreeAllocation` will be defined as a property of the `TeamType` - (ref: https://github.com/flowforge/flowforge/issues/782#issuecomment-1241742266)

It defaults to 0 - as it stands we haven't decided what values we will apply (if any). The purpose of that part of this PR is to ensure we have this capability.


I will add some additional tests for it - just requires a bit of juggling around of things.

**Comment by knolleary** [comment]
I have now added the final test for deviceFreeAllocation.

**Pull request merged**